juju_spell/commands/update_packages.py

```diff
@@ -0,0 +1,195 @@
+                    unit.raw_output = result
+
+    def parse_result(self, result: str):
+        """Parse result.
+
+        Parses the result and creates PackageUpdateResult structure.
+        """
+        lines = result.splitlines()
+        packages: List[PackageUpdateResult] = []
+        for line in lines:
+            # Inst libdrm2 [2.4.110-1ubuntu1] (2.4.113-2~ubuntu0.22.04.1 Ubuntu:22.04/jammy-updates [amd64]) # noqa
+            to_version = ""
+            from_version = ""
+            name = ""
+            if line.startswith("Inst "):
+                _, name, from_version, to_version, *others = line.split(" ")
+
+            # Unpacking software-properties-common (0.99.9.11) over (0.99.9.10)
+            elif line.startswith("Unpacking"):
+                _, name, to_version, _, from_version, *others = line.split(" ")
+
+            to_version = to_version.strip("()[]")
+            from_version = from_version.strip("()[]")
+            name = name.strip(" ")
+            if from_version != "" and to_version != "" and name != "":
+                packages.append(
+                    PackageUpdateResult(
+                        package=name, from_version=from_version, to_version=to_version
+                    )
+                )
+
+        return packages
+
+    def get_update_command(self, app: Application, dry_run: bool):
+        """Generate command according to flags."""
+        template = UPDATE_TEMPLATE + ("--dry-run" if dry_run else "")
+        if app.dist_upgrade:
+            return template.format(install="dist-upgrade", packages="")
+        else:
+            app_list = [a.package for a in app.packages_to_update]
+            package_list = " ".join(app_list)
+            return template.format(install="install", packages=package_list)
+
+    def set_apps_to_update(self, model: Model, updates: Updates, dry_run: bool):
+        """Set units to applications.
+
+        Finds the matching applications and set the units of these applications as a
+        List[UpdateResult] to application.
+        """
+        for update in updates.applications:
+            command = self.get_update_command(app=update, dry_run=dry_run)
+            for app, app_status in model.applications.items():
+                if re.match(update.name_expr, app):
+                    unit_updates = [
+                        UnitUpdateResult(unit=u.name, command=command)
+                        for u in app_status.units
+                    ]
+                    update.results.append(
+                        UpdateResult(application=app, units=unit_updates)
+                    )
+
+    def set_success_flags(
+        self, unit: UnitUpdateResult, expected: List[PackageToUpdate]
+    ):
+        """Set success flag for each unit."""
+        expected_set = set([(e.version, e.package) for e in expected])
+        real_set = [(p.to_version, p.package) for p in unit.packages]
```

Comment:
You missed one `set` and you do not need to use list comprehension here. `set` will handle the generator
```suggestion
        expected_set = set((e.version, e.package) for e in expected)
        real_set = set((p.to_version, p.package) for p in unit.packages)
```